lib/SPIRVProducerPass.cpp

```diff
@@ -2850,6 +2859,230 @@ void SPIRVProducerPass::GenerateResourceVars(Module &) {
   }
 }
 
+namespace {
+
+bool isScalarType(Type *type) {
+  return type->isIntegerTy() || type->isFloatTy();
+}
+
+unsigned structAlignment(StructType *type,
+                         std::function<unsigned(Type *)> alignFn) {
+  unsigned maxAlign = 1;
+  for (unsigned i = 0; i < type->getStructNumElements(); i++) {
+    unsigned align = alignFn(type->getStructElementType(i));
+    maxAlign = std::max(align, maxAlign);
+  }
+  return maxAlign;
+}
+
+unsigned scalarAlignment(Type *type) {
+  // A scalar of size N has a scalar alignment of N.
+  if (isScalarType(type)) {
+    return type->getScalarSizeInBits() / 8;
+  }
+
+  // A vector or matrix type has a scalar alignment equal to that of its
+  // component type.
+  if (type->isVectorTy()) {
+    return scalarAlignment(type->getVectorElementType());
+  }
+
+  // An array type has a scalar alignment equal to that of its element type.
+  if (type->isArrayTy()) {
+    return scalarAlignment(type->getArrayElementType());
+  }
+
+  // A structure has a scalar alignment equal to the largest scalar alignment of
+  // any of its members.
+  if (type->isStructTy()) {
+    return structAlignment(cast<StructType>(type), scalarAlignment);
+  }
+
+  llvm_unreachable("Unsupported type");
+}
+
+unsigned baseAlignment(Type *type) {
+  // A scalar has a base alignment equal to its scalar alignment.
+  if (isScalarType(type)) {
+    return scalarAlignment(type);
+  }
+
+  if (type->isVectorTy()) {
+    unsigned numElems = type->getVectorNumElements();
+
+    // A two-component vector has a base alignment equal to twice its scalar
+    // alignment.
+    if (numElems == 2) {
+      return 2 * scalarAlignment(type);
+    }
+    // A three- or four-component vector has a base alignment equal to four
+    // times its scalar alignment.
+    if ((numElems == 3) || (numElems == 4)) {
+      return 4 * scalarAlignment(type);
+    }
+  }
+
+  // An array has a base alignment equal to the base alignment of its element
+  // type.
+  if (type->isArrayTy()) {
+    return baseAlignment(type->getArrayElementType());
+  }
+
+  // A structure has a base alignment equal to the largest base alignment of any
+  // of its members.
+  if (type->isStructTy()) {
+    return structAlignment(cast<StructType>(type), baseAlignment);
+  }
+
+  // TODO A row-major matrix of C columns has a base alignment equal to the base
+  // alignment of a vector of C matrix components.
+  // TODO A column-major matrix has a base alignment equal to the base alignment
+  // of the matrix column type.
+
+  llvm_unreachable("Unsupported type");
+}
+
+unsigned roundUpToMultipleOf(unsigned num, unsigned multiple) {
```

Comment:
This is effectively MathExtras.h `alignTo` (with Skew = 0). I'd prefer reusing that to adding another definition.

Reply:
Thanks! I was sure there was already a function for this somewhere.